File: TwoWayDatabinding/app/src/main/java/com/enpassio/twowaydatabinding/utils/AppExecutors.java
package com.enpassio.twowaydatabinding.utils;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class AppExecutors {

    // For Singleton instantiation
    private static final Object LOCK = new Object();
    private static AppExecutors sInstance;
    private final Executor diskIO;

    private AppExecutors(Executor diskIO) {
        this.diskIO = diskIO;
    }

    public static AppExecutors getInstance() {
        if (sInstance == null) {
            synchronized (LOCK) {
                sInstance = new AppExecutors(Executors.newSingleThreadExecutor());
            }
        }
        return sInstance;
    }

    public Executor diskIO() {
        return diskIO;
    }
}

File: TwoWayDatabindingKotlin/app/src/main/java/com/enpassion/twowaydatabindingkotlin/utils/AppExecutors.kt
package com.enpassion.twowaydatabindingkotlin.utils

import java.util.concurrent.Executor
import java.util.concurrent.Executors

class AppExecutors private constructor(private val diskIO: Executor) {

    fun diskIO(): Executor {
        return diskIO
    }

    companion object {

        // For Singleton instantiation
        private val LOCK = Any()

        private var sInstance: AppExecutors? = null

        fun getInstance(): AppExecutors {
            return sInstance ?: synchronized(LOCK) {
                sInstance ?: AppExecutors(Executors.newSingleThreadExecutor()).also { sInstance = it }
            }
        }
    }
}